config/common.go

```diff
@@ -141,3 +174,95 @@ func (cs *commonStore) validate(cfg *model.Config) error {
 
 	return nil
 }
+
+// removeEnvOverrides returns a new config without the current environment overrides.
+// If a config variable has an environment override, that variable is set to the value that was
+// read from the store.
+func (cs *commonStore) removeEnvOverrides(cfg *model.Config) *model.Config {
+	// When saving, iterate through the environmentOverrides map and check:
+	// foreach envOverrides, if config == loadedConfigNoEnv, then the environment override matched the original setting. No change
+	// foreach envOverrides, if config != loadedConfigNoEnv, then:
+	//    a) if config == loadedConfig, persist loadedConfigNoEnv (we don't want to persist the envSetSetting).
+	//    b) if config != loadedConfig, persist config (the user has subsequently changed the setting after the envOverride was applied)
+	//       NOTE: b cannot happen at the moment (from docs: "if a setting is configured through an environment variable,
+	//       modifying it in the System Console is disabled").
+	// Therefore, if config != loadedConfigNoEnv, always persist the loadedConfigNoEnv value
+
+	paths := getPaths(cs.environmentOverrides)
+	newCfg := cfg.Clone()
+	for _, path := range paths {
+		currentVal := getVal(newCfg, path)
+		loadedVal := getVal(cs.loadedConfigNoEnv, path)
+		if currentVal.Interface() != loadedVal.Interface() {
+			setVal(newCfg, path, loadedVal.Interface())
+		}
+	}
+	return newCfg
+}
+
+// getPaths is helper function for removeEnvOverrides
+func getPaths(m map[string]interface{}) [][]string {
+	return getPathsRec(m, nil, nil)
+}
+
+// getPathsRec is helper function for removeENvOverrides
+func getPathsRec(src interface{}, curPath []string, allPaths [][]string) [][]string {
```

Comment:
nit: rather than passing in `allPaths` onto which we accumulate (but never read inside this method), I believe this function could just return its view of `allPaths` and let the parent call accumulate, making for a simpler recursive model:

```go
func getPathsRec(src interface{}, curPath []string) [][]string {
	if reflect.ValueOf(src).Kind() == reflect.Map {
		paths := [][]string{}
		for k, v := range src.(map[string]interface{}) {
			paths = append(paths, getPathsRec(v, append(curPath, k))
		}
		return paths
	}

	return [][]string{curPath}
}
```

Reply:
Secondly, I think the `reflect.ValueOf` might be better captured as a cast attempt instead, given we cast to an even more strict type anyway:
```go
	if srcMap, ok := src.(map[string]interface{}); ok {
		paths := [][]string{}
		for k, v := range srcMap {
			paths = append(paths, getPathsRec(v, append(curPath, k))
		}
		return paths
	}
```

Reply:
Excellent rewriting of getPathsRec (as I mentioned over coffee).  And the cast attempt is much cleaner, thanks!